File: wsip/lab3/c/dwumian.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>

typedef unsigned long long num_t;

// Oblicza symbola newtona = n ! / (k ! * (n - k) !)
// Przy uzyciu tablicy i trojkata pascala
num_t binomial_coefficient(int n, int k)
{
    // wadliwe dane
    if (n < k)
        return 0;

    // inicjalizacja tablicy, z kazdym elementem = 0
    num_t coeffs[k + 1];
    memset(coeffs, 0, sizeof(num_t)*(k + 1));
    coeffs[0] = 1;

    // 'i' poziom w trojkacie, temp 
    num_t i = 1, temp, prev_copy;

    for(num_t i = 1; i <= n; i++)
    {
        // zmienna pomocnicza, by zachowac wartosc z poprzedniego elementu
        prev_copy = 1;
        const num_t min = i < k ? i : k;

        for(num_t j = 1; j <= min; j++)
        {
            // coeffs[j] += coeffs[j-1] dziala tylko dla pierwszego elementu,
            // dlatego potrzebny jest 'temp' i 'prev_copy', by zachowac stan z przed 
            // dodawania
            temp = coeffs[j];
            coeffs[j] = coeffs[j] + prev_copy;
            prev_copy = temp;
        }

        // jesli i <= k, to przypisujemy 1 (koniec trojkata)
        if (i <= k)
            coeffs[i] = 1;
    }

    return coeffs[k];
}

int main(int argc, char* argv[])
{
    int n = 0, k = 0;
    printf("Podaj 2 liczby: ");
    scanf("%d %d", &n, &k);
    printf("Symbol Newtona (%d %d) = %lu\n", n, k, binomial_coefficient(n, k));
    return 0;
}

File: wsip/lab3/py/dwumian.py
import numpy as np


def binomial_coefficient(n: int, k: int) -> np.int64:
    """
    Oblicza wartość dwumianu Newtona (n k) = n! / (k! * (n - k)!)
    Przy użyciu trójkąta Pascala, zwraca wartość dwumianu dla pary (n, k).
    """

    if n < k:
        raise ValueError(f'Niepoprawne dane, {n} < {k}')
    
    # Inicjalizacja tablicy dwuwymiarowej z 0, n + 1 wierszy, k + 1 kolumn,
    # Ponieważ chcemy obliczyć wartość dwumianu dla pary (n, k)
    coeffs = np.zeros((k + 1), dtype=np.int64)
    coeffs[0] = 1

    for i in range(1, n + 1):        
        # min(i, k) + 1, ucinamy wartości, które nie mają sensu
        # zatem dla danego poziomu n, obliczamy wartości dla k od 1 do min(i, k)
        # a nie cały zakres tego poziomu ( = i + 1)
        prev_copy = 1
        for j in range(1, min(i, k) + 1):
            temp = coeffs[j]
            coeffs[j] += prev_copy
            prev_copy = temp
        
        if i <= k:
            coeffs[i] = 1

    return coeffs[k]



def main():
    n: int = int(input('Podaj n: '))
    k: int = int(input('Podaj k: '))

    try:
        result = binomial_coefficient(n, k)
        print(f'({n} {k}): {result}')
    except ValueError as e:
        print(e)


if __name__ == '__main__':
    main()